nix/build-dapp-package.nix

```diff
@@ -0,0 +1,75 @@
+{ pkgs }:
+
+let
+  remappings = xs:
+    builtins.foldl' pkgs.lib.mergeAttrs {}
+      (builtins.map
+        (x: {
+          "${x.name}/" = "${x}/dapp/${x.name}/src/";
+         } // x.remappings)
+         xs);
+  libPaths = xs:
+    builtins.foldl' pkgs.lib.mergeAttrs {}
+      (builtins.map
+        (x: {
+          "${x.name}" = "${x}/dapp/${x.name}/src";
+         } // x.libPaths)
+         xs);
+in
+  {
+    name
+  , src
+  , deps ? []
+  , solc ? pkgs.solc
+  , test-hevm ? pkgs.hevm
+  , solcFlags ? ""
+  , doCheck ? true
+  , ... }@args:
+    pkgs.stdenv.mkDerivation ( rec {
+      inherit doCheck;
+
+      buildInputs = [ test-hevm solc ];
+      passthru = {
+        remappings = remappings deps;
+        libPaths = libPaths deps;
+      };
+
+      REMAPPINGS =
+        pkgs.lib.mapAttrsToList
+          (k: v: k + "=" + v)
+          passthru.remappings;
+
+      LIBSCRIPT =
+        pkgs.lib.mapAttrsToList
+          (k: v: ''
+            ln -s ${v} lib/${k}
+          '')
+          passthru.libPaths;
+
+      buildPhase = args.buildPhase or ''
+        jsonopts=--combined-json=abi,bin,bin-runtime,srcmap,srcmap-runtime,ast
+
+        mkdir -p out
+
+        find "$src" -name '*.sol' | while read -r x; do
+          (set -x; ${pkgs.solc}/bin/solc --overwrite $REMAPPINGS --allow-paths $src $solcFlags --abi --bin --bin-runtime -o out "$x")
+          json_file=out/''${x##*/}.json
+          (set -x; ${pkgs.solc}/bin/solc $REMAPPINGS --allow-paths $src $solcFlags $jsonopts "$x" >"$json_file")
+        done
+
+        mkdir lib
+        echo "$LIBSCRIPT" > setup.sh
+        source setup.sh
+        export DAPP_LIB=lib
+      '';
+
+      checkPhase = args.checkPhase or ''
+        hevm dapp-test
+      '';
+      installPhase = args.installPhase or ''
+        mkdir -p $out/dapp/$name
+        cp -r $src $out/dapp/$name/src
+        cp -r lib $out/dapp/$name/lib
```

Comment:
It could be interesting to do a multiple outputs derivation instead of this, so that they would result in different store paths.

This could mean that if e.g. a dependency doesn't change, that store path remains unchanged.

Reply:
deduplication?

Reply:
Right, maybe for `src` we would get some deduplication in some (rare?) cases. Not for `lib` though, there are only symlinks in there.

I think it's pretty cool aesthetically so we should probably do it anyway